src/libraries/xvnode/src/xsniffer.cpp

```diff
@@ -0,0 +1,373 @@
+    common::xnode_type_t type{m_vnode->vnetwork_driver()->type()};
+    bool disable_broadcasts{false};
+    xdbg("[xtop_vnode::get_roles_data] node type : %s", common::to_string(type).c_str());
+
+    if (common::has<common::xnode_type_t::consensus_auditor>(type)) {
+        xdbg("[xtop_vnode::get_roles_data] add all sharding contracts' rcs");
+        type = common::xnode_type_t::consensus_validator;
+        disable_broadcasts = true;
+    }
+
+    auto const & system_contract_deployment_data = m_system_contract_manager->deployment_data();
+    for (auto const & contract_data_pair : system_contract_deployment_data) {
+        auto const & contract_address = top::get<common::xaccount_address_t const>(contract_data_pair);
+        auto const & contract_data = top::get<contract_runtime::system::xcontract_deployment_data_t>(contract_data_pair);
+
+        if (!common::has(type, contract_data.node_type)) {
+            continue;
+        }
+        if (!m_config_map.count(contract_address)) {
+            m_config_map.insert(std::make_pair(contract_address, xrole_config_t{contract_data, std::map<common::xaccount_address_t, uint64_t>()}));
+        }
+        if (disable_broadcasts) {
+            m_config_map[contract_address].role_data.broadcast_config.zone = contract_runtime::xsniff_broadcast_type_t::invalid;
+        }
+    }
+}
+
+xsniffer_config_t xtop_sniffer::sniff_config() const {
+    xsniffer_config_t config;
+    for (auto const & data_pair : m_config_map) {
+        auto const & data = top::get<xrole_config_t>(data_pair);
+        auto const & sniff_type = data.role_data.sniff_type;
+        if (contract_runtime::has<contract_runtime::xsniff_type_t::broadcast>(sniff_type)) {
+            auto sniff_block_type = data.role_data.broadcast_config.type;
+            assert(sniff_block_type != contract_runtime::xsniff_block_type_t::invalid);
+            config.insert(std::make_pair(xsniffer_event_type_t::broadcast, xsniffer_event_config_t{sniff_block_type, std::bind(&xtop_sniffer::sniff_broadcast, this, std::placeholders::_1)}));
+        }
+        if (contract_runtime::has<contract_runtime::xsniff_type_t::timer>(sniff_type)) {
+            config.insert(std::make_pair(xsniffer_event_type_t::timer,
+                                         xsniffer_event_config_t{xsniffer_block_type_t::all_block, std::bind(&xtop_sniffer::sniff_timer, this, std::placeholders::_1)}));
+        }
+        if (contract_runtime::has<contract_runtime::xsniff_type_t::block>(sniff_type)) {
+            auto sniff_block_type = data.role_data.block_config.type;
+            assert(sniff_block_type != contract_runtime::xsniff_block_type_t::invalid);
+            config.insert(std::make_pair(xsniffer_event_type_t::block, xsniffer_event_config_t{sniff_block_type, std::bind(&xtop_sniffer::sniff_block, this, std::placeholders::_1)}));
+        }
+    }
+
+    return config;
+}
+
+bool xtop_sniffer::sniff_broadcast(xobject_ptr_t<base::xvblock_t> const & vblock) const {
```

Comment:
this API's prototype should follow our coding convention:
`void xtop_sniffer::sniff_broadcast(xobject_ptr_t<base::xvblock_t> const & block, std::error_code & ec) const;`